tests/components/screenlogic/test_init.py

```diff
@@ -241,3 +251,43 @@ async def test_entity_migration_data(
 
     entity_not_migrated = entity_registry.async_get(old_eid)
     assert entity_not_migrated == original_entity
+
+
+@pytest.mark.parametrize("data_set", [DATA_FULL_NO_GPM, DATA_FULL_CHEM])
+async def test_platform_setup(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry, data_set: dict
+) -> None:
+    """Test platform setup."""
+    stub_connect = lambda *args, **kwargs: stub_async_connect(data_set, *args, **kwargs)
+
+    def add_ents(ents: list[Any]) -> None:
+        pass
+
+    with patch(
+        GATEWAY_DISCOVERY_IMPORT_PATH,
+        return_value={},
+    ), patch.multiple(
+        ScreenLogicGateway,
+        async_connect=stub_connect,
+        is_connected=True,
+        _async_connected_request=DEFAULT,
+    ):
+        connect_info = await async_get_connect_info(hass, mock_config_entry)
+        gateway = ScreenLogicGateway()
+        await gateway.async_connect(**connect_info)
+
+        mock_coordinator = Mock(
+            spec=ScreenlogicDataUpdateCoordinator,
+            config_entry=mock_config_entry,
+            gateway=gateway,
+            hass=hass,
+        )
+        hass.data.setdefault(DOMAIN, {})[mock_config_entry.entry_id] = mock_coordinator
+        for platform in PLATFORMS:
```

Comment:
Don't access integration details like hass.data or platforms in the tests. Set up the integration via the config entry while patching the library appropriately.

Reply:
What I was trying to test for was if the platform setup correctly handled missing data for the expected sensor(s). The thought was to eventually use the JSON fixtures to test against a variety of different equipment configurations the integration could encounter.

The problem I was running into with this test in particular was that I couldn't figure out a way catch the exception from as far up as is typically desired by setting up the integration via the config entry.

The exception seemed to be caught in `EntityPlatform._async_setup_platform()` and converted to returning `False`, returned up through `EntityComponent` to `ConfigEntry.async_setup()` where a false result for `await component.async_setup_entry(hass, self)` seems to only be used to set the `state` on the config entry to `ConfigEntryState.SETUP_ERROR`.  I tried asserting on the config entry's state, but it was always 'loaded'.

The only thing I could think to do to catch the exception was to mock/bootstrap the integration's `async_setup()` (what you see with accessing hass.data) and directly call the integration's platform `async_setup_entry()`s.

I'm sure there are reasons, but it is a little frustrating to seemingly not be able to assert the success of platform setup from `hass.config_entries.async_forward_entry_setups()`. Maybe there's an avenue that I'm missing?

Reply:
I'm not following about the exception to catch. There should be no exception, right? We can assert that the expected entities have created a state in the state machine.

Reply:
If the platform setup correctly skips setting up entities where there will be missing data, then correct there should be no exception. 

I was trying to make the test more (maybe too much) generic in that it could catch missing data that we didn't think about by catching the exception that is raised when the expected data is not found. Sort of a "Canary in the coal mine". But you're right; no exception != good test. In this case the test could pass but the validity of the entities is still uncertain. We should be asserting on the actual expected state of things.  It was just that there are so many permutations of that to test, and that are not currently known.

This all may be moot, though. The most recent version of the api library doesn't have any data excluded so we shouldn't run into this anyways. Still interested in expanding my knowledge of HA though.

Reply:
If there's an exception during platform setup no entities of that platform will be added. So it's enough to check that one entity per platform is added to know that platform setup was successful while creating the missing data case for each platform by patching the library data.

Reply:
b7a2cbd